Introduce Location and Label enum. Replace old ingredients with Location.

lukasdenk opened the pull request:
Fixes #48

Philipp000 commented on src/entities.py:
```diff
@@ -92,7 +97,305 @@ def __hash__(self) -> int:
+        "13": {COCOA_CONTAINING_GREASE},  # type: ignore
+        "14": {GELATIN},  # type: ignore
+        "99": {ALCOHOL},  # type: ignore
+        # meatless is not an ingredient
+        "f": {},  # type: ignore
+        # vegan is not an ingredient
+        "v": {},  # type: ignore
```
Even though meatless and vegan are no ingredients, I think it is still reasonable to save this information

lukasdenk replied:
True. But maybe somewhere else?

Philipp000 replied:
I think, it. would be better, to make the category Ingredients more general, such as Labels. 
For example GQB or MSC are also no real ingredients.

Alternatively one could also use the dish_type for this, but I think I would opt for the more general approach, of putting all information into the "renamed" Ingredients.

lukasdenk commented:
@COM8 Are you sure that no project uses the ingredients?
![telegram_mensa_bot](https://user-images.githubusercontent.com/63459921/143853868-d1c6fc26-29ac-445a-aa90-c798744eaca9.jpeg)

lukasdenk commented:
**PR is ready now.**

## Summary
- Introduced `Location` enum
- Introduced `Label` enum. `Label` contains ingredients and other labels like `VEGAN`, etc..
- `parse.sh` now automatically creates Json files from the `Location` and `Label` enum.
- Updated `README`. Changed section _Ingredients_ to section _Labels_. Links to the generated label json.

COM8 commented:
> I did not add a DishType enum since I do not think that the DishType will be displayed by a client. @COM8 What do you think?

I agree with not adding a `DishType` enum since dish types change rather frequently at Mensa Garching. Almost every month there is a new one.
I'm using and displaying the dish type in the Windows TCA.

> I did not add symbols to the ingredients, like @Philipp000 suggested in Enums for Ingredients and Location #48. Maybe this could be done in another PR.

Sounds reasonable to keep the size of the PR down.

COM8 commented:
Then please convert the PR from a draft to "Ready for review".

COM8 commented on src/test/assets/mediziner-mensa/reference/week_2018_44.json:
```diff
@@ -0,0 +1,482 @@
+                "guests": null
+              },
+              "labels": [
+                "Label.ANTIOXIDANTS",
```
I don't think this is expected output here, right?

lukasdenk replied:
Has been like this before. But I'll add the price to staff and guests now, too. According to the [testfile](https://github.com/TUM-Dev/eat-api/blob/495c1530c8590a9edb5518782632eb6bb2f9a052/src/test/assets/mediziner-mensa/for-generation/week_2018_44.txt) this should be fine. There does not seem to be a special price for staff or guests.  
However, I can't verify whether this is the case now, since I can't find the canteen's menu on there webpage. Probably the canteen is closed.

lukasdenk replied:
I just added a new issue for the problem. See #67.

Philipp000 reviewed (commented):
What do we need the Files in resources for? So `resources/lables.json` and `resources/locations.json`?
I think this is kind of duplicate, as we have now enums + these json files?

As far as I'm concerned, having only the enums should be sufficient? The json, files are then generated in the API. Otherwise, a change has to be always done at both places (enum + resources folder)?

lukasdenk commented on src/entities.py:
```diff
@@ -95,193 +95,241 @@ def __hash__(self) -> int:
-        # dishes contain the same ingredients
-        else:
-            self._values_lookup(split_values, None)
+class Site:
```
I renamed it to Location. Since previously a canteen was sometimes refered to as canteen and sometimes as location, I now changed it so that a canteen is consistently refered to as canteen.

lukasdenk commented on src/entities.py:
```diff
@@ -95,193 +95,241 @@ def __hash__(self) -> int:
+    def __init__(self, german_text: str):
+        self.german_text = german_text
+
+    GLUTEN = "Gluten"
```
Good idea. I applied your proposal except that I replaced "de" by a newly introduced Enum `Language`.

Philipp000 commented on src/main.py:
```diff
@@ -80,18 +82,15 @@ def main():
     if args.locations:
-        with open("canteens.json", "r", encoding="utf-8") as canteens:
-            print(json.dumps(json.load(canteens)))
+        with open("canteens.json", "r", encoding="utf-8") as locations:
```
Shouldn't this be generated through the new enums?

lukasdenk replied:
:+1:

Philipp000 commented on src/entities.py:
```diff
@@ -95,193 +95,241 @@ def __hash__(self) -> int:
+
+class Location(Enum):
+    # Some of the locations do not use the general Studentenwerk system and therefore do not have a url_id.
+    def __init__(self, long_name: str, site: Site, url_id: int):
```
We should add an canteen_id, to have the same functionality as before.

lukasdenk replied:
Which functionality do you mean?

Philipp000 replied:
As far as I'm concerned, the `canteen_id` is used, to identify the folder containing the dishes for the specific canteen.
So for example the "Mensa Garching" has the canteen_id `mensa-garching`, and in the gh-pages (so the API) you can find it in the path with `mensa-garching`: https://github.com/TUM-Dev/eat-api/tree/gh-pages/mensa-garching

Further, this is also required for the website, or other programs to function properly.
This has in mind, that the `canteens.json` should be removed and instead generated through the enum.

lukasdenk replied:
Ok, I'll add it.

Philipp000 commented on scripts/parse.sh:
```diff
@@ -28,7 +28,7 @@ python3 scripts/combine.py
 # Copy canteens.json in the output directory:
-echo "Coppying canteens..."
+echo "Copying canteens..."
 cp src/canteens.json $OUT_DIR
```
Should also be generated through the new enums.
Additionally, we should delete the `canteens.json`, to not have the information stored double, and to prevent errors, when changes are made to these.

lukasdenk replied:
True. But don't we need the old file for backwards compability?

Philipp000 replied:
I'd say the format of the file is alright, so ideally I would propose, that you use the same format, and maybe (for now) also copy your newly generated file to the `canteens.json` file?

Anyway, we could also let the old file there for now, however, I think we should not copy it anymore, as there shouldn't be a canteens.json file in the master branch anymore.

lukasdenk replied:
Well, I know copy it from ENUM_DIR/canteens.json

Philipp000 commented on resources/locations.json:
```diff
@@ -0,0 +1,212 @@
+[
+  {
+    "name": "MENSA_ARCISSTR",
+    "long_name": "Mensa Arcisstra\u00dfe",
```
Why is it necessary, to have a name and a long name? 
To me it does not seem, as if the name will ever be displayed?

Shouldn't we try to reproduce the same format as in the canteens.json? So the `long_name` should be `name`, and the mensa-id would be needed.

COM8 replied:
Agree. This is a relic from a previous implementation.

lukasdenk replied:
You are basically right. But `Canteen` inherits from `Enum`. `Enum` however has the field `name` which is the name of the enum, like `MENSA_ARCISSTR` in the above example. In your approach, the `name` attribute in the Json object would mean something completely different than the `name` attribute in the `Canteen` object.  


Instead, I think, one should use the `name` attribute as the id. I only introduced the `canteen_id` now to keep backward compability.

Philipp000 replied:
Okay alight, however in my opinion the output of the API should be determined by the usefulness and information of the output, and not by the internal mechanics of the programming language used to generate it.
Additionally, it feels wrong to me, to use a field called `name` as id.

Just imagine, we switch to another mechanic, and use some other class/configuration file, etc. to generate the locations we would have to change the API again.

Additionally, if we want to use the new IDs/names (which i think is not necessary by now) we also have to rename all the folders of the current API. However, I think as long as the API format makes sense, we should keep it that way, and not change it due to some internal mechanics.

COM8 replied:
Also making such significant changes will break almost all implementations.

lukasdenk replied:
You are right about the API. I applied your suggestions now.

Philipp000 commented on resources/locations.json:
```diff
@@ -0,0 +1,212 @@
+    "url_id": 418
+  },
+  {
+    "name": "STUBISTRO_GROSSHADERN",
```
I mentioned the issue with the `ß` in this canteen-id in issue #68.
When we change it here, we also need to move the files in the gh-pages as I described in the issue

Philipp000 reviewed (changes requested):
I think, we should completely replace the canteens.json with the new enum.
Here the canteen-id is missing, as far as I'm concerned.

lukasdenk commented on resources/labels.json:
```diff
@@ -0,0 +1,214 @@
+[
+  {
+    "name": "GLUTEN",
+    "german_text": "Gluten"
```
see comment on src.entities l 204-207

lukasdenk commented:
> What do we need the Files in resources for? So `resources/lables.json` and `resources/locations.json`? I think this is kind of duplicate, as we have now enums + these json files?
> 
> As far as I'm concerned, having only the enums should be sufficient? The json, files are then generated in the API. Otherwise, a change has to be always done at both places (enum + resources folder)?

sorry, I used them for testing. Forgot to delete them

Philipp000 commented on README.md:
```diff
@@ -38,71 +28,8 @@ Simple static API for the canteens of the [Studentenwerk München](http://www.st
-</p>
-</details>
+## Label list - previously _ingredients_:
+See [here](https://tum-dev.github.io/eat-api/dist/enums/labels.json).
```
I think, the path must be: https://tum-dev.github.io/eat-api/enums/labels.json

So without the `dist` subfolder. As only everything in the dist folder is pushed to the gh-pages branch.

lukasdenk replied:
:+1:

Philipp000 commented on README.md:
````diff
@@ -137,10 +64,10 @@ This results in this file being usually half the size of the `all.json` file.
 
 ```
-https://tum-dev.github.io/eat-api/canteens.json
+https://tum-dev.github.io/eat-api/dist/enums/canteens.json
````
Here also without `dist/` subpath

lukasdenk replied:
:+1:

Philipp000 commented on scripts/parse.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-LOC_LIST=( "mensa-arcisstr" "mensa-garching" "mensa-leopoldstr" "mensa-lothstr" \
+CANTEEN_LIST=( "mensa-arcisstr" "mensa-garching" "mensa-leopoldstr" "mensa-lothstr" \
```
Can we replace this list, to read the available canteens from the python enum?

lukasdenk replied:
I like your idea. But I think using the shell script to parse the enums to a list and then call `main.py` for each canteen would not be the proper way. Instead, we should add an option directly to `main.py` to parse all canteens at once. I have created an issue for that. See #70.

Philipp000 commented on scripts/parse.sh:
```diff
@@ -28,15 +28,20 @@ python3 scripts/combine.py
-    python3 src/main.py -p "$loc" --openmensa "./dist/$loc"
+for canteen in "${openmensa_list[@]}"; do
+    echo "Parsing openmensa menus for: " "$canteen"
+    python3 src/main.py -p "$canteen" --openmensa "./dist/$canteen"
```
I think, instead of hard coding dist, you should use the OUT_DIR variable

lukasdenk replied:
:+1:

Philipp000 commented on scripts/parse.sh:
```diff
@@ -28,15 +28,20 @@ python3 scripts/combine.py
+    python3 src/main.py -p "$canteen" --openmensa "./dist/$canteen"
 done
 
+ENUM_JSON_PATH="./dist/enums"
```
Here again, the OUT_DIR variable is probably better

lukasdenk replied:
ok

COM8 commented on README.md:
```diff
@@ -166,23 +93,23 @@ The JSON files are produced by the tool shown in this repository. Hence, it is e
   --openmensa PATH      directory for OpenMensa XML output (date parameter
                         will be ignored if this argument is used)
-  -l, --locations       prints all available locations formated as JSON
+  -l, --canteens       prints all available canteens formated as JSON
```
That’s no good interface. But I do not know how we could call it better since we have the clash with the `—combine` option.

lukasdenk replied:
just remove the abbrevation, I guess

COM8 replied:
That’s actually a really good idea 👍🏻

COM8 commented on src/menu_parser.py:
```diff
@@ -1,28 +1,34 @@
     Abstract menu parser class.
     """
 
+    canteens: Set[Canteen] = NotImplemented
```
Why do you assign a `NotImplemented` here?

lukasdenk replied:
I saw it as a solution to provide an abstract attribute. But now I have read that one does not need to provide a value for an abstract class. I'll change it.

COM8 commented on src/menu_parser.py:
```diff
@@ -1,28 +1,34 @@
     """
 
+    canteens: Set[Canteen] = NotImplemented
+    _label_lookup: Dict[str, Set[Label]] = NotImplemented
```
Same as above.

Philipp000 commented on src/menu_parser.py:
```diff
@@ -57,14 +96,64 @@ def __init__(self, price):
+        "13": {Label.COCOA_CONTAINING_GREASE},
+        "14": {Label.GELATIN},
+        "99": {Label.ALCOHOL},
+        # meatless is not a label
```
shouldn't this be vegetarian now?

lukasdenk replied:
True

Philipp000 commented on src/menu_parser.py:
```diff
@@ -157,92 +246,44 @@ def __get_self_service_prices(
+    base_url: str = "http://www.studentenwerk-muenchen.de/mensa/speiseplan/speiseplan_{canteen_id}_-de.html"
     base_url_with_date: str = (
-        "http://www.studentenwerk-muenchen.de/mensa/speiseplan/speiseplan_{date}_{location_id}_-de.html"
+        "http://www.studentenwerk-muenchen.de/mensa/speiseplan/speiseplan_{date}_{canteen_id}_-de.html"
```
Does this also work with the canteen_id (which should be e.g. `mensa-garching`)?
As far as I tested it, it doesn't, but the `url_id` or here previously `location_id` is needed?

lukasdenk replied:
oh, sorry. The url_id is needed

lukasdenk replied:
I'll change it

lukasdenk commented:
**Ready for merge now.** I have built in your proposals now.

COM8 commented:
```bash
python src/main.py -p mensa-garching
```
results in:

```bash
Traceback (most recent call last):
  File "/tmp/eat-api/src/main.py", line 144, in <module>
    main()
  File "/tmp/eat-api/src/main.py", line 140, in main
    print(weeks[calendar_week])
  File "/tmp/eat-api/src/entities.py", line 424, in __repr__
    week_str += f"\n {day}"
  File "/tmp/eat-api/src/entities.py", line 390, in __repr__
    return str(self.menu_date) + ": " + str(self.dishes)
  File "/tmp/eat-api/src/entities.py", line 356, in __repr__
    return f"{self.name} {str(sorted(self.labels))}: {str(self.prices)}"
  File "/tmp/eat-api/src/entities.py", line 289, in __lt__
    return self.name < other.value
TypeError: '<' not supported between instances of 'str' and 'dict'
```

COM8 commented on src/enum_json_creator.py:
```diff
@@ -0,0 +1,30 @@
+
+
+if __name__ == "__main__":
+    base_directory = "dist/enums"
```
It fails if the directory structure does not exist. Example:
```bash
python3 src/enum_json_creator.py
Traceback (most recent call last):
  File "/tmp/eat-api/src/enum_json_creator.py", line 30, in <module>
    write_enum_as_api_representation_to_file(base_directory, value, key)
  File "/tmp/eat-api/src/enum_json_creator.py", line 21, in write_enum_as_api_representation_to_file
    file_util.write(os.path.join(base_dir, filename), enum_to_api_representation_dict(list(enum_type)))
  File "/tmp/eat-api/src/utils/file_util.py", line 37, in write
    with open(path, "w", encoding="utf-8") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'dist/enums/canteens.json'
```

It would be better to either create the directories recursively or print a warning.

lukasdenk replied:
:+1:

COM8 commented on src/enum_json_creator.py:
```diff
@@ -0,0 +1,30 @@
+if __name__ == "__main__":
+    base_directory = "dist/enums"
+    if len(sys.argv) > 1:
+        base_directory = sys.argv[1]
```
Please add some minimal documentation for this script in the `README.md`.

lukasdenk replied:
:+1:

COM8 reviewed (changes requested):
Only a few changes left and then we are ready to go 🎉.

lukasdenk commented:
Fixed the bug with parsing garching and applied the suggested changes.

COM8 reviewed (approved):
Looks good now. Now please rebase once more and we are ready to go.

lukasdenk commented:
@COM8 Done :+1:
